### PR TITLE
fix available modules

### LINKED ISSUE #11524: `import pytorch_lightning` fails with `ValueError: transformers.models.auto.__spec__ is None`
## 🐛 Bug

A ValueError occurs whenever PyTorch Lightning is imported after Huggingface's Transformer. Importing Transformer after PyTorch does not lead to any errors, however. 

### To Reproduce
Tested in local anaconda environment:
import this:
```
from transformers import T5ForConditionalGeneration, T5Tokenizer
import pytorch_lightning
```
It seems that this issue doesn't exist on Google Colab.

### Expected behavior
```
ValueError: transformers.models.auto.__spec__ is None
```

### Environment
* CUDA:
        - GPU:
                - TITAN X (Pascal)
                - TITAN X (Pascal)
                - TITAN X (Pascal)
                - TITAN X (Pascal)
        - available:         True
        - version:           10.1
* Packages:
        - numpy:             1.22.1
        - pyTorch_debug:     False
        - pyTorch_version:   1.8.1+cu101
        - pytorch-lightning: 1.5.8
        - tqdm:              4.61.2
        - transformers: 4.15.0
* System:
        - OS:                Linux
        - architecture:
                - 64bit
                - ELF
        - processor:         x86_64
        - python:            3.9.5
        - version:           #124-Ubuntu SMP Fri Nov 10 18:29:59 UTC 2017

### PR DESCRIPTION
**Pull request by Borda**
## What does this PR do?

Fixes #11524
Original fix was taken from https://github.com/PyTorchLightning/metrics/pull/772

## Before submitting

- [x] Was this **discussed/approved** via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your **PR does only one thing**, instead of bundling different changes together?
- [x] Did you make sure to **update the documentation** with your changes? (if necessary)
- [x] Did you write any **new necessary tests**? (not for typos and docs)
- [x] Did you verify new and **existing tests pass** locally with your changes?
- [ ] Did you list all the **breaking changes** introduced by this pull request?
- [ ] Did you **update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)**? (not for typos, docs, test updates, or internal minor changes/refactorings)

## PR review

Anyone in the community is welcome to review the PR.
Before you start reviewing make sure you have read [Review guidelines](https://github.com/PyTorchLightning/pytorch-lightning/wiki/Review-guidelines). In short, see the following bullet-list:

- [x] Is this pull request ready for review? (if not, please submit in draft mode)
- [x] Check that all items from **Before submitting** are resolved
- [x] Make sure the title is self-explanatory and the description concisely explains the PR
- [x] Add labels and milestones (and optionally projects) to the PR so it can be classified

## Did you have fun?

Make sure you had fun coding 🙃

cc @borda

### CODE REVIEW

File: pytorch_lightning/utilities/imports.py [hunk 1]
```diff
@@ -25,18 +25,46 @@
...
+        # Python 3.7+
+        return False
+
+
 def _module_available(module_path: str) -> bool:
-    """Check if a path is available in your environment.
+    """Check if a module path is available in your environment.
 
     >>> _module_available('os')
     True
```
Comment (akihironitta) [inline pytorch_lightning/utilities/imports.py:50]: For checking package availability, one has to use `_package_available` above, right?
```suggestion
```

Comment (Borda) [reply pytorch_lightning/utilities/imports.py:50]: It shall be optional

File: pytorch_lightning/utilities/imports.py [hunk 2]
```diff
@@ -25,18 +25,46 @@
...
+    False
+    """
+    try:
+        return find_spec(package_name) is not None
+    except AttributeError:
+        # Python 3.6
+        return False
+    except (ImportError, ModuleNotFoundError):
+        # Python 3.7+
+        return False
```
Comment (carmocca) [inline pytorch_lightning/utilities/imports.py:43]:
```suggestion
    except ModuleNotFoundError:
        return False
```

File: pytorch_lightning/utilities/imports.py [hunk 3]
```diff
@@ -25,18 +25,46 @@
...
+
+
 def _module_available(module_path: str) -> bool:
-    """Check if a path is available in your environment.
+    """Check if a module path is available in your environment.
 
     >>> _module_available('os')
     True
+    >>> _module_available('os.bla')
+    False
```
Comment (carmocca) [inline pytorch_lightning/utilities/imports.py:52]:
```suggestion
```

There's already the test case right below

File: pytorch_lightning/utilities/imports.py [hunk 4]
```diff
@@ -72,26 +100,26 @@ def _compare_version(package: str, op: Callable, version: str, use_base_version:
...
 _APEX_AVAILABLE = _module_available("apex.amp")
-_DEEPSPEED_AVAILABLE = _module_available("deepspeed")
+_DEEPSPEED_AVAILABLE = _package_available("deepspeed")
 _FAIRSCALE_AVAILABLE = not _IS_WINDOWS and _module_available("fairscale.nn")
 _FAIRSCALE_OSS_FP16_BROADCAST_AVAILABLE = _FAIRSCALE_AVAILABLE and _compare_version("fairscale", operator.ge, "0.3.3")
 _FAIRSCALE_FULLY_SHARDED_AVAILABLE = _FAIRSCALE_AVAILABLE and _compare_version("fairscale", operator.ge, "0.3.4")
 _GROUP_AVAILABLE = not _IS_WINDOWS and _module_available("torch.distributed.group")
-_HOROVOD_AVAILABLE = _module_available("horovod.torch")
-_HYDRA_AVAILABLE = _module_available("hydra")
+_HOROVOD_AVAILABLE = _package_available("horovod.torch")
```
Comment (carmocca) [inline pytorch_lightning/utilities/imports.py:108]: Should this be module?

```suggestion
_HOROVOD_AVAILABLE = _module_available("horovod.torch")
```

File: pytorch_lightning/utilities/imports.py (hunk 3 above)
Comment (Borda) [reply pytorch_lightning/utilities/imports.py:52]: yes but `os` is valid...

Comment (carmocca) [reply pytorch_lightning/utilities/imports.py:48]: Then remove

```python
    >>> _module_available('bla.bla')
    False
```

?

**Review by carmocca — APPROVED** [review]

**Review by rohitgr7 — APPROVED** [review]

File: pytorch_lightning/utilities/imports.py (hunk 3 above)
Comment (Borda) [reply pytorch_lightning/utilities/imports.py:48]: not sure what you mean by move, `_module_available('bla.bla')` is testing completely bad path...

**Review by s-rog — APPROVED** [review]

**Review by tchaton — APPROVED** [review]
LGTM !

**Comment by tchaton** [comment]
@Borda I think the `requires` decorator from Lightning Flash by @ethanwharris would also be a great addtion.

**Comment by Borda** [comment]
> @Borda I think the `requires` decorator from Lightning Flash by @ethanwharris would also be a great addtion.

if whole function or class needs a certain dependency, then yes, but in Flash, in several places, we have a switch inside function/method body... anyway a wrapper is welcome but it is outside of scope of this PR :rabbit:

**Review by justusschock — APPROVED** [review]

**Review by akihironitta — APPROVED** [review]

**Pull request merged**